### Données de 18/02/2019 00:00 jusqu'à 18/02/2019 16:22 de toutes les stations

In [33]:
#Charger les biblio
import numpy as np
import pandas as pd
from pandas import read_csv
import matplotlib.pyplot as plt
from matplotlib import cm
import seaborn as sns
import numpy as np
import string 
from subprocess import check_output
import seaborn as sns
from pylab import *
%matplotlib inline
sns.set()

In [153]:
# Lire la table avec 'header'
df = pd.read_csv('Hmvl2.csv', sep=';', engine='python')

In [154]:
# Remplacer les NaN par des zeros dans trame_hmvl
df.trame_hmvl.fillna('0', inplace=True)

In [155]:
# Reindexer la dataframe
df.index = range(len(df))

In [156]:
#changeaons les noms de colonnes
df.columns = ['code', 'id','equipement_id', 'date','trame']

In [128]:
df.head()

code    id  equipement_id                     date  \
0                   MBS  7357           7357  2019-02-18 00:00:00.146   
1                   MPA  7356           7356  2019-02-18 00:00:00.146   
2                   MPH  7355           7355  2019-02-18 00:00:00.146   
3  SFB-TPCCEN_RDTSCB001  7850           7850  2019-02-18 00:00:01.374   
4  SMA-INTBAU_RDTSCB001  7859           7859  2019-02-18 00:00:01.374   

                                               trame  
0  0543110903705431109037256930750412569307504125...  
1                                                  @  
2                 058520920440585209204405852092044@  
3                                                  @  
4                                                  @

In [157]:
# On enlève les @
df["trame"] = df["trame"].str.replace("@", "")

In [158]:
# On ajoute une colonne Count pour compter chacune
df['Count'] = df['trame'].str.len()

In [159]:
df.head()

code    id  equipement_id                     date  \
0                   MBS  7357           7357  2019-02-18 00:00:00.146   
1                   MPA  7356           7356  2019-02-18 00:00:00.146   
2                   MPH  7355           7355  2019-02-18 00:00:00.146   
3  SFB-TPCCEN_RDTSCB001  7850           7850  2019-02-18 00:00:01.374   
4  SMA-INTBAU_RDTSCB001  7859           7859  2019-02-18 00:00:01.374   

                                               trame  Count  
0  0543110903705431109037256930750412569307504125...     55  
1                                                         0  
2                  058520920440585209204405852092044     33  
3                                                         0  
4                                                         0

In [132]:
df.Count.max()/11

46.0

In [160]:
# lignes differentes de zero
df = df[df.Count != 0]

In [161]:
# lignes differentes de 1
df = df[df.Count != 1]

In [135]:
df.groupby("Count").count().head()

code      id  equipement_id    date   trame
Count                                               
5           1       1              1       1       1
11     149327  149327         149327  149327  149327
12        245     245            245     245     245
16          1       1              1       1       1
22     106444  106444         106444  106444  106444

In [136]:
dff = df.groupby("Count").count()

In [141]:
# La somme cumulé
dff['CUMS'] = dff['id'].cumsum(axis = 0)

In [145]:
dff[60:75]

code     id  equipement_id   date  trame  Count  CUMSUM    CUMS
Count                                                                  
107        1      1              1      1      1    107    3958  827858
108        1      1              1      1      1    108    4066  827859
110    33234  33234          33234  33234  33234    110    4176  861093
111      149    149            149    149    149    111    4287  861242
112      130    130            130    130    130    112    4399  861372
113       89     89             89     89     89    113    4512  861461
114       43     43             43     43     43    114    4626  861504
115       15     15             15     15     15    115    4741  861519
116        1      1              1      1      1    116    4857  861520
117        1      1              1      1      1    117    4974  861521
119        1      1              1      1      1    119    5093  861522
120       66     66             66     66     66    120    5213  861588
121    22843  22843          22843  22843  22843    121    5334  884431
122      132    132            132    132    132    122    5456  884563
123      128    128            128    128    128    123    5579  884691

Le nombre totale de caractère qui représente $100\%$ est de $929852$.

$884431$ représente $95.11\%$ de données qui correspond à la valeur $121$ au champs 'Count'.

On considère alors, pour ces données, le nombre maximal de caractère est de $11$.

Donc il faut créer 11 nouveaux champs qui contient les 11 caractères pour chaque véhicules.

In [162]:
# Supprimons la colonne
df = df.drop(['Count'], axis=1)

In [163]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10']

In [164]:
# Ajouter les colonnes, en séparant la colonne 'trame' 
for i in range(len(liste)):
    df[liste[i]] = df["trame"].str.slice(11*i,11*(i+1))

In [165]:
df.head(3)

code    id  equipement_id                     date  \
0  MBS  7357           7357  2019-02-18 00:00:00.146   
2  MPH  7355           7355  2019-02-18 00:00:00.146   
6  M1b  6383           6383  2019-02-18 00:00:02.069   

                                               trame           x0  \
0  0543110903705431109037256930750412569307504125...  05431109037   
2                  058520920440585209204405852092044  05852092044   
6                                        15736085041  15736085041   

            x1           x2           x3           x4 x5 x6 x7 x8 x9 x10  
0  05431109037  25693075041  25693075041  25896106044                     
2  05852092044  05852092044                                               
6

In [168]:
# Supprimer les colonnes xj
df0 = df.drop(['x3','x1','x2','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
# df1
df1 = df.drop(['x3','x0','x2','x4','x5','x6','x7','x8', 'x9', 'x10',  'trame'], axis=1)
# Supprimer les colonnes xj
df2 = df.drop(['x3','x1','x0','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
df3 = df.drop(['x0','x1','x2','x4','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
# Supprimer les colonnes xj
df4 = df.drop(['x0','x1','x2','x3','x5','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
# Supprimer les colonnes xj
df5 = df.drop(['x0','x1','x2','x4','x3','x6','x7','x8', 'x9', 'x10', 'trame'], axis=1)
# Supprimer les colonnes xj
df6 = df.drop(['x0','x1','x2','x4','x5','x3','x7','x8', 'x9', 'x10', 'trame'], axis=1)
# Supprimer les colonnes xj
df7 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x8', 'x9', 'x10', 'trame'], axis=1)
# Supprimer les colonnes xj
df8 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x9', 'x10', 'trame'], axis=1)
df9 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x8', 'x10',  'trame'], axis=1)

df10 = df.drop(['x0','x1','x2','x4','x5','x6','x3','x7', 'x9', 'x8','trame'], axis=1)


In [169]:
liste = ["x0","x1","x2","x3","x4","x5","x6","x7","x8","x9",'x10']
liste2 = [df0,df1,df2,df3,df4,df5,df6,df7,df8,df9,df10]

    # Séparer la colonne xi en 4 colonnes  
for i, f in zip(range( len(liste)), liste2):
    f["voie"] = f[liste[i]].str.slice(0,1)
    f["horodatage"] = f[liste[i]].str.slice(1,5)
    f["vitesse"] = f[liste[i]].str.slice(5,8)
    f["longueur"] = f[liste[i]].str.slice(8,11)

In [170]:
df0.head()

code    id  equipement_id                     date           x0 voie  \
0  MBS  7357           7357  2019-02-18 00:00:00.146  05431109037    0   
2  MPH  7355           7355  2019-02-18 00:00:00.146  05852092044    0   
6  M1b  6383           6383  2019-02-18 00:00:02.069  15736085041    1   
8  M1d  6385           6385  2019-02-18 00:00:02.069  15004114040    1   
9  M1e  6386           6386  2019-02-18 00:00:02.069  12780093040    1   

  horodatage vitesse longueur  
0       5431     109      037  
2       5852     092      044  
6       5736     085      041  
8       5004     114      040  
9       2780     093      040

In [171]:
# rename columns
for i,f in zip(range( len(liste)), liste2):
    
    f.rename(columns={liste[i]:'x'}, inplace=True)

In [172]:
df2.head()

code    id  equipement_id                     date            x voie  \
0  MBS  7357           7357  2019-02-18 00:00:00.146  25693075041    2   
2  MPH  7355           7355  2019-02-18 00:00:00.146  05852092044    0   
6  M1b  6383           6383  2019-02-18 00:00:02.069                     
8  M1d  6385           6385  2019-02-18 00:00:02.069                     
9  M1e  6386           6386  2019-02-18 00:00:02.069                     

  horodatage vitesse longueur  
0       5693     075      041  
2       5852     092      044  
6                              
8                              
9

In [173]:
# Concatener les data
result = pd.concat(liste2)

In [175]:
result.head()

code    id  equipement_id                     date            x voie  \
0  MBS  7357           7357  2019-02-18 00:00:00.146  05431109037    0   
2  MPH  7355           7355  2019-02-18 00:00:00.146  05852092044    0   
6  M1b  6383           6383  2019-02-18 00:00:02.069  15736085041    1   
8  M1d  6385           6385  2019-02-18 00:00:02.069  15004114040    1   
9  M1e  6386           6386  2019-02-18 00:00:02.069  12780093040    1   

  horodatage vitesse longueur  
0       5431     109      037  
2       5852     092      044  
6       5736     085      041  
8       5004     114      040  
9       2780     093      040

In [176]:
# Reindexer la dataframe
result.index = range(len(result))

In [179]:
# Supprimons la colonne
df = result.drop(['x'], axis=1)

In [180]:
c = ['horodatage', 'vitesse']
for i in c:
    df[i] = pd.to_numeric(df[i] ,errors='coerce')

In [181]:
# Eliminer les lignes vides
df = df.dropna(axis=0, how='any', thresh=None, subset=None, inplace=False)

In [182]:
len(result)-len(df)

5759785

In [184]:
#Exporter la table
df.to_csv('df_5sc_to.csv', index = False, encoding = 'utf-8')

 Voici le lien vers la table de données de de 18/02/2019 00:00 jusqu'à 18/02/2019 16:22 de toutes les satations :
 
 https://github.com/Oumayma13/Data_5sc
 
 Sous le nom de 'df_5sc_to'.